hazelcast-client/src/test/java/com/hazelcast/client/heartbeat/ClientHeartbeatTest.java

```diff
@@ -0,0 +1,147 @@
+        hazelcastFactory.terminateAll();
+    }
+
+    @Test
+    public void testHeartbeatStoppedEvent() throws InterruptedException {
+        HazelcastInstance instance = hazelcastFactory.newHazelcastInstance();
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_TIMEOUT, "3000");
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_INTERVAL, "500");
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
+
+        HazelcastClientInstanceImpl clientImpl = getHazelcastClientInstanceImpl(client);
+        ClientConnectionManager connectionManager = clientImpl.getConnectionManager();
+        final CountDownLatch countDownLatch = new CountDownLatch(1);
+        connectionManager.addConnectionHeartbeatListener(new ConnectionHeartbeatListener() {
+            @Override
+            public void heartBeatStarted(Connection connection) {
+            }
+
+            @Override
+            public void heartBeatStopped(Connection connection) {
+                countDownLatch.countDown();
+            }
+        });
+
+        blockMessagesFromInstance(instance, client);
+        assertOpenEventually(countDownLatch);
+    }
+
+    @Test(expected = TargetDisconnectedException.class)
+    public void testInvocation_whenHeartbeatStopped() throws InterruptedException {
+        hazelcastFactory.newHazelcastInstance();
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_TIMEOUT, "3000");
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_INTERVAL, "500");
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
+        HazelcastInstance instance2 = hazelcastFactory.newHazelcastInstance();
+
+        String keyOwnedByInstance2 = generateKeyOwnedBy(instance2);
+        IMap map = client.getMap(randomString());
+        map.put(keyOwnedByInstance2, randomString());
+        blockMessagesFromInstance(instance2, client);
+        map.put(keyOwnedByInstance2, randomString());
+    }
+
+    @Test
+    public void testAsyncInvocation_whenHeartbeatStopped() throws InterruptedException {
+        hazelcastFactory.newHazelcastInstance();
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_TIMEOUT, "3000");
+        clientConfig.setProperty(ClientProperty.HEARTBEAT_INTERVAL, "500");
+        HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
+        HazelcastInstance instance2 = hazelcastFactory.newHazelcastInstance();
+
+        IMap map = client.getMap(randomString());
+        String keyOwnedByInstance2 = generateKeyOwnedBy(instance2);
+        map.put(keyOwnedByInstance2, randomString());
+        blockMessagesFromInstance(instance2, client);
+        ClientDelegatingFuture future = (ClientDelegatingFuture) map.putAsync(keyOwnedByInstance2, randomString());
```

Comment:
I am blocking when response turning back.  

How can I make sure `getAsync` not finished when I call to `blockMessagesFromInstance` method after `getAsync` ? Before executing next line, it could be the case that response already got back.